src/Microsoft.DotNet.XHarness.InstrumentationBase.Xunit/AndroidInstrumentationBase.cs

```diff
@@ -0,0 +1,198 @@
+            {
+                bundleArguments.Add(ResultsFileArgumentName, "TestResults.xml");
+            }
+
+            Start();
+        }
+
+        public override async void OnStart()
+        {
+            base.OnStart();
+
+            var bundle = new Bundle();
+
+            var entryPoint = new TestsEntryPoint(bundleArguments);
+            entryPoint.TestsCompleted += (sender, results) =>
+            {
+                var message =
+                    $"Tests run: {results.ExecutedTests} " +
+                    $"Passed: {results.PassedTests} " +
+                    $"Inconclusive: {results.InconclusiveTests} " +
+                    $"Failed: {results.FailedTests} " +
+                    $"Ignored: {results.SkippedTests}";
+                bundle.PutString(SummaryMessageArgumentName, message);
+                
+                bundle.PutLong(ExitCodeArgumentName, results.FailedTests == 0 ? 0 : 1);
+            };
+
+            entryPoint.Tests = Tests;
+
+            await entryPoint.RunAsync();
+
+            if (File.Exists(entryPoint.TestsResultsFinalPath))
+            {
+                bundle.PutString(ResultsPathArgumentName, entryPoint.TestsResultsFinalPath);
+            }
+            
+            if (bundle.GetLong(ExitCodeArgumentName, -1) == -1)
+            {
+                bundle.PutLong(ExitCodeArgumentName, 1);
+            }
+
+            Finish(Result.Ok, bundle);
+        }
+
+        class TestsEntryPoint : AndroidApplicationEntryPoint
+        {
+            readonly string resultsPath;
+            readonly string includeMethod;
+            readonly string excludeCategories;
+            readonly string excludeMethod;
+            readonly string includeClass;
+            readonly string excludeClass;
+
+            public TestsEntryPoint(Dictionary<string, string> arguments)
+            {
+                var root = Application.Context.GetExternalFilesDir(null)?.AbsolutePath;
+
+                var docsDir = Path.Combine(root, "Documents");
+
+                if (!Directory.Exists(docsDir))
+                {
+                    Directory.CreateDirectory(docsDir);
+                }
+
+                resultsPath = Path.Combine(docsDir, arguments[ResultsFileArgumentName]);
+                includeMethod = arguments[IncludeMethodArgumentName];
+                excludeCategories = arguments[ExcludeCategoriesArgumentName];
+                excludeMethod = arguments[ExcludeMethodArgumentName];
+                includeClass = arguments[IncludeClassArgumentName];
+                excludeClass = arguments[ExcludeClassArgumentName];
+            }
+
+            protected override bool LogExcludedTests => true;
+
+            public override TextWriter Logger => null;
+
+            public override string TestsResultsFinalPath => resultsPath;
+
+            protected override int? MaxParallelThreads => System.Environment.ProcessorCount;
+
+            protected override IDevice Device => null;
```

Comment:
Same, nullable references, we should make sure that this will not result in a NRE, if it does not, lets add IDevice?

Reply:
I checked, it could be null and don't be cause of exceptions